homeassistant/components/aurora_abb_powerone/__init__.py

```diff
@@ -45,3 +49,58 @@ async def async_unload_entry(hass: HomeAssistant, entry: ConfigEntry) -> bool:
         hass.data[DOMAIN].pop(entry.entry_id)
 
     return unload_ok
+
+
+class AuroraAbbDataUpdateCoordinator(DataUpdateCoordinator):
+    """Class to manage fetching AuroraAbbPowerone data."""
+
+    def __init__(self, hass: HomeAssistant, comport: str, address: int) -> None:
+        """Initialize the data update coordinator."""
+        self.available_prev: bool = False
+        self.available: bool = False
+        self.client = AuroraSerialClient(address, comport, parity="N", timeout=1)
+        super().__init__(hass, _LOGGER, name=DOMAIN, update_interval=SCAN_INTERVAL)
+
+    async def _async_update_data(self) -> dict[str, float]:
+        """Fetch new state data for the sensor.
+
+        This is the only method that should fetch new data for Home Assistant.
+        """
+        data: dict[str, float] = {}
+        try:
+            self.available_prev = self.available
+            self.client.connect()
+
+            # read ADC channel 3 (grid power output)
+            power_watts = self.client.measure(3, True)
+            data["instantaneouspower"] = round(power_watts, 1)
+
+            temperature_c = self.client.measure(21)
+            data["temp"] = round(temperature_c, 1)
+
+            energy_wh = self.client.cumulated_energy(5)
+            data["totalenergy"] = round(energy_wh / 1000, 2)
+            self.available = True
+            self.data = data
+
+        except AuroraTimeoutError:
+            self.data = {}
+            self.available = False
```

Comment:
I have done some experiments with this, and I'm not happy with the result when raising `UpdateFailed`.  

The problem is that `UpdateFailed` logs an error every time, but going offline during the night is not an error.  I don't want to fill the user's logs with errors that are impossible to prevent.  By explicitly detecting online/offline I can manage what generates an error and also only show log messages when the device comes back online, even though it is trying 100s of times in the night.

Reply:
Maybe it changed in all those months, but I just checked the code of the DataUpdateCoordinator and there is a mechanism for this:
```python3
        except UpdateFailed as err:
            self.last_exception = err
            if self.last_update_success:
                if log_failures:
                    self.logger.error("Error fetching %s data: %s", self.name, err)
                self.last_update_success = False
```
Which is doing exactly what you're doing right now

Reply:
I get what you are saying, but I'm trying to do something very specific, which is to catch the case of a timeout error, then display a single warning message on the first instance.

I think the automatic route only gives me the possibility to suppress all failure logging which isn't what I want.

In addition, this particular device is sometimes unreliable particularly in low light, so I'm planning a follow up PR that does some auto-retrying, see https://github.com/davet2001/home-assistant/commit/f7455239300b2ead06bbfa45e480613f358b71ea and https://github.com/home-assistant/core/issues/82983

Reply:
I am not talking about the `log_failures` but about the last_update_success. If you trigger an UpdateFailed, it will log an error, and then it will set `last_update_success` to False, which is checked before logging.

So in practice, if UpdateFailed is raised 5 times, 1 error is written. It won't log it again, until it actually does an update without a raised error (so it succeeded).

Reply:
Yes, they are doing very similar things.  I did a quick experiment, changing the code to raise `UpdateFailed` on timeout.  It works as you describe.  But here is the log output:

Existing code:
```
2023-08-25 11:30:04.467 WARNING (SyncWorker_2) [homeassistant.components.aurora_abb_powerone] Communication with aurora_abb_powerone lost
2023-08-25 11:35:06.608 INFO (SyncWorker_4) [homeassistant.components.aurora_abb_powerone] Communication with aurora_abb_powerone back online
```

Using UpdateFailed to manage offline/online messages:
```
2023-08-25 11:26:35.351 ERROR (MainThread) [homeassistant.components.aurora_abb_powerone] Error fetching aurora_abb_powerone data: No response after 3 tries
2023-08-25 11:28:13.427 INFO (MainThread) [homeassistant.components.aurora_abb_powerone] Fetching aurora_abb_powerone data recovered
```

They are almost identical, but the key point:  _going offline due to sunset is not an error_.  So I want a way of this occurring without raising anything red in the logs (only info/warning).  I can't see any other way of doing this.

Reply:
Okay you got my point, now I know we're on the same wavelength.

Is there a way in the code you can differentiate on if it's because of a failure or if it's because of sunset?

Reply:
Unfortunately none that I know of, unplugging the cable behaves the same as sunset (both cause a timeout).

Reply:
Right, I still think this is a difficult situation. imo, posting something in the logs doesn't have to be seen as something bad, just an FYI. I think we can just use the normal way like every integration does, but just have to teach the user a bit more about how it works via the documentation or the error raised. You could edit the raised error like, "Lost connection to inverter, could it be dark?". We have way more integrations that just log every time the service is responds with a 500 while the next call just succeeds.

I think it would be good to have a third opinion on this matter. Like I get where you come from, but I am just wondering what's the right solution here.

Reply:
@joostlek I don't agree. If there are warning or error prints users will create issues and complain about it. In this case, what could be considered is to log with info level since debug prints are usually very spammy.

I think the PR is good to go as is though, adjusting to info level can be done in a follow-up.